### PR TITLE
Fields typing improvements

### PR DESCRIPTION
**Pull request by bellini666**
This was built on top of https://github.com/sbdchd/django-types/pull/63, so merging it should only keep this PR's change

Note that even though the change looks big, it actually is pretty small, it is just that all fields need to define their own `__new__` (or `__init__` if their signature is different from the base field), and it is all the same, the difference is the classname and the generic field. I also added some documentation [here](https://github.com/sbdchd/django-types/pull/65/files#diff-876a5d116ef70c86039fbb8498623871324fa93fc1ae51c7c09ba2a0f7612e99R47) explaining the usage.

Fix https://github.com/sbdchd/django-types/issues/34

### CODE REVIEW

**Comment by bellini666** [comment]
This was rebased on top of the main branch. Please note again that the tests in here will not pass since they change `Field` in a way that makes all subclasses have to implement that, and the changes to `ArrayField` and all `RelatedField` subclasses are on #63 and #64. Those being merged will make the tests in here pass as well.

Also note that this fixes all base fields typing issues, including things like https://github.com/sbdchd/django-types/pull/69 which was just merged.

**Comment by bellini666** [comment]
@sbdchd just rebased this on top of [#73](https://github.com/sbdchd/django-types/pull/63) and simplified some code, also adding some more testings.

Please tell me if there's anything missing for this to be merged.

**Comment by sbdchd** [comment]
yeah looks god, once it's fixed in relation to master it should be good to merge

**Comment by sbdchd** [comment]
ditto for the other PR

**Comment by bellini666** [comment]
Hey @sbdchd , thanks!

Don't know why this had conflicts. Each of those PRs were based on the other, so merging one should not make the next ones conflict.

Well, either way they are all rebased now! The order for merging should be:

1) https://github.com/sbdchd/django-types/pull/63
2) This PR (which is based on that one)
3) https://github.com/sbdchd/django-types/pull/66 (which is based on this one)

Because of that, if you prefer you can just merge this to merge #63 and this together or even merge #66 which will be the same as merging all 3 together.

File: tests/pyright/main.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+f = Foo()
+reveal_type(f.char_with_choices)
+f.char_with_choices == "c"
+f.char_with_choices == "a"
+f.char_with_choices = "c"
+f.char_with_choices = "b"
+"""
+)
+assert _char_test == [
```
Comment (sbdchd) [inline tests/pyright/main.py:110]: thoughts on wrapping this in pytest? might help with the error messages and usability

Comment (bellini666) [reply tests/pyright/main.py:110]: Sure! Will do that right now!

I thought that maybe you might not want to add pytest as a dependency, but for sure it can help with error messages and usability.

**Comment by bellini666** [comment]
@sbdchd just changed the pyright tests to use pytest instead

**Comment by bellini666** [comment]
Hey @sbdchd ,

Don't know if you didn't see this because I didn't rerequest your review. Either way, pinging you so you can take a look at this again.

File: django-stubs/contrib/postgres/fields/array.pyi [hunk 2]
```diff
@@ -26,71 +22,105 @@ _T = TypeVar("_T", bound=Optional[List[Any]])
...
+    Field[Union[_T, Combinable], _T],
 ):
 
     empty_strings_allowed: bool = ...
     default_error_messages: Any = ...
     base_field: Any = ...
     size: Any = ...
     default_validators: Any = ...
-    from_db_value: Any = ...
+    from_dbevalue: Any = ...
```
Comment (sbdchd) [inline django-stubs/contrib/postgres/fields/array.pyi:33]: typo?

File: tests/trout/models.py [hunk 3]
```diff
@@ -263,7 +289,7 @@ def main() -> None:
     process_non_nullable(for_sure_c.integer)
 
     # Django way to duplicate an instance
-    comment.id = None
```
Comment (sbdchd) [inline tests/trout/models.py:266]: why the change from `.id` to `.pk`?

**Review by sbdchd — COMMENTED** [review]
Looks good!

**Comment by sbdchd** [comment]
Thanks!
